### PR TITLE
[FIX] Bad layout for button Published

### PR DESCRIPTION
**Pull request by kujiu**
Bad layout for button "Published" on website.

What's the good behavior ?

* Published button is unstylized.

### CODE REVIEW

File: addons/web/static/src/scss/fonts.scss [hunk 1]
```diff
@@ -35,3 +35,12 @@ $lato-font-path: '../fonts/lato';
 @include lato-font-pair('Reg', 400);
 @include lato-font-pair('Bol', 700);
 @include lato-font-pair('Bla', 900);
+
+.fa {
+    font: normal normal normal normal normal;
```
Comment (qsm-odoo) [inline addons/web/static/src/scss/fonts.scss]: Invalid property value.
`font-family` cannot be set to `normal`

File: addons/web/static/src/scss/ui.scss [hunk 2]
```diff
@@ -13,3 +13,8 @@
 [aria-hidden="true"], [aria-hidden="1"] {
     display: none!important;
 }
+
+.btn-unstylized {
```
Comment (qsm-odoo) [inline addons/web/static/src/scss/ui.scss]: Basically, this is equal to the `btn bg-transparent` combo with BS4... but well, I suppose it could be ok.

**Comment by kujiu** [comment]
> What if we want a button with an icon on the left and an icon on the right ?

<span class="fa fa-chevron-left">Previous</span><span class="fa fa-suffix-chevron-right">Next</span>

I did all we need to do that.

> What if we want a button with a specific icon for mobile and another one for desktop ?

Are you sure the label will be the same ?

> What if we want an icon only for desktop ?

Title and aria-label are wanted on desktop and not on mobile, so all the code is different.

> It seems that the correct approach is defined here: https://fontawesome.com/how-to-use/on-the-web/other-topics/accessibility
>  Basically, hiding all decorative fonts would have been easier and would not break anything I think ?

The sr-only class has been defined by Bootstrap for the screen readers. But it's not implemented on a majority of screen readers, so it's useless and doesn't respect the WAI documents of the W3C. I've seen this way, and none of my tests work (with Orca on Linux, with NVDA on Windows, with VoiceOver on MacOS X and with Talkback on Android). And in any case, the screen reader does not read decorative fonts as they use specific UTF8 codes, so the aria-hidden doesn't change anything. There's also an other problem : <i> does not mean image or icon but a portion of text like a dialog between two persons or a sentence in foreign language. So normally, you have to add role="img" to say it's not a text but an icon.

An other problem of this approach concerns the way to do checks. It's really difficult to think about describing anything, and nobody do that. Automatic checks permit to be sure all is labelled and we need to link a description to the element.

**Comment by qsm-odoo** [comment]
Hello,

I was talking about those possible cases:
```xml
<button class="btn btn-primary">
    <i class="fa fa-arrow-up"></i> Go ! <i class="fa fa-arrow-up"></i>
</button>
```
```xml
<button class="btn btn-primary">
    <i class="fa fa-arrow-left hidden-xs"></i><i class="fa fa-long-arrow-left fa-4x hidden-sm hidden-md hidden-lg"></i> Go back
</button>
```
```xml
<button class="btn btn-primary">
    <i class="fa fa-plus hidden-xs"></i> Add
</button>
```
In the last one for example, your changes will hide the whole button while we only wanted to hide the extra decorative font. However, the solution of just adding `aria-hidden` on the 'plus' decorative font should be enough ?

I don't understand your explication about `aria-hidden`. How could a screen reader not understand this one ? I don't understand the need of linking useless images to anything for a screen reader, it seems to me that having less elements/references is in fact better ?

Also, the point of the `sr-only` class is not creating references, but replacing semantic icon or content by screen readers content. Is this really not working on a lot of screen readers ? Bootstrap implemented the class years ago and I cannot find someone complaining about it, this is even recommended by some non-bootstrap users. Also, in case you would want to create references, what about using `aria-label` / `aria-labelledby` ?

**Comment by kujiu** [comment]
For these cases, the best solution is just aria-label on button. All the content of the button is replaced by aria-label for the screen reader, like it:

`<button class="btn btn-primary" aria-label="Go !">
    <i class="fa fa-arrow-up"></i> Go ! <i class="fa fa-arrow-up"></i>
</button>`

> In the last one for example, your changes will hide the whole button while we only wanted to hide the extra decorative font. However, the solution of just adding aria-hidden on the 'plus' decorative font should be enough ?

The screen reader doesn't recognize characters of the decorative font: there are in a specific UTF-8 range so the screen-reader doesn't read them. So, hiding the element is useless.

> I don't understand your explication about aria-hidden. How could a screen reader not understand this one ? I don't understand the need of linking useless images to anything for a screen reader, it seems to me that having less elements/references is in fact better ?

When browsing with a screen reader, we navigate in a simili-DOM by searching patterns and by browsing a tree. If we have more specific patterns, the search is more efficient. For example, if we have an image before each actions to do in a text, we can extract the list of actions by selecting images. (This is very useful to search trains on the SNCB website for example. Sincerely, it's already a mess without aria-hidden, I hope they never think about it !). And when we are on the good element, we can navigate around. This is like searching a specific tag with jquery, then go to the sibling, and in the child, etc. The tree is simplier than the DOM, and queries are done by shortcuts. But you can imagine a navigation like our JS tours but without clicking, just with the keyboard.

The big problem with aria-hidden is really simple: it's not used correctly on a majority of websites and developpers who knows how to use it are very rare. So, a lot of blind people configure their screen readers to don't care about aria-hidden. This is the only way to browse the majority of websites correctly.

> Also, the point of the sr-only class is not creating references, but replacing semantic icon or content by screen readers content. Is this really not working on a lot of screen readers ? Bootstrap implemented the class years ago and I cannot find someone complaining about it, this is even recommended by some non-bootstrap users. Also, in case you would want to create references, what about using aria-label / aria-labelledby ?

aria-label and aria-labelledby is the WAI-compliant way to do it. sr-only is just a hack from BS done before HTML5. You can do anything done by sr-only by using aria-label. Like this `<span aria-label="Please do that" role="status"/>` or this `<i class="fa fa-something" aria-label="Something" role="img"/>`. No need to replace the <i> element. (The role is really important as aria-label does not work on span or on i without it.)

And we need also to force developpers to describe icons, to keep accessibility over the future versions.

**Comment by qsm-odoo** [comment]
> If we have more specific patterns, the search is more efficient. For example, if we have an image before each actions to do in a text, we can extract the list of actions by selecting images.

But if the image is just a decoration inside a button with a text, isn't the fact that the whole element is a button with a text enough to determine it is an action ?

**Comment by kujiu** [comment]
Yes, it's good if it's in a button if there is a description on the button (this was not always the case).

**Comment by qsm-odoo** [comment]
Well, if I understood everything right, this case:
```xml
<button class="btn btn-primary"><i class="fa fa-home"/> Go back home</button>
```
can be left untouched, as the icon is useless and the aria-hidden on it is not needed and the button is screen-readable.

And this case:
```xml
<button class="btn btn-primary"><i class="fa fa-home"/></button>
```
should be replaced by
```xml
<button class="btn btn-primary fa fa-home" aria-label="Go back home"/>
```

In any case, patching the FontAwesome lib is not a solution. You took care of applying the font on the `::before` element but what about `fa-2x`, `fa-3x`, `fa-spin`, `fa-rotate-x`, ... ? I agree I don't see the reason why FontAwesome is applying the font this way, but we have to live with it (and there may be a reason). Acting against it will also make a potential switch to FA 5.0.0 harder.

**Comment by kujiu** [comment]
Ok for that, but how do you test that a developper set a description in all cases when using a fa-* class with just some xpath expressions ?

**Comment by qsm-odoo** [comment]
Not sure to understand the question... :/

**Comment by kujiu** [comment]
Can you valid just the presentation of the button ? An other PR comes soon for the other changes.

**Pull request merged**